### PR TITLE
[SelectionDAG] Introducing the SelectionDAG pattern matching framework

### PR DESCRIPTION
**Pull request by mshockwave**
Akin to `llvm::PatternMatch` and `llvm::MIPatternMatch`, the `llvm::SDPatternMatch` introduced in this patch provides a DSL-alike framework to  match SDValue / SDNode with a more succinct syntax.

### CODE REVIEW

**Comment by mshockwave** [comment]
Note: I only added the most basic and hopefully the most common patterns in this patch. Feel free to propose other patterns that you thought are important.

File: llvm/include/llvm/CodeGen/SDPatternMatch.h [hunk 1]
```diff
@@ -0,0 +1,612 @@
...
+
+inline Opcode_match m_Opc(unsigned Opcode) { return Opcode_match(Opcode); }
+
+template <unsigned NumUses, typename Pattern> struct NUses_match {
+  Pattern P;
+
+  explicit NUses_match(const Pattern &P) : P(P) {}
+
+  bool match(const SelectionDAG *DAG, SDValue N) {
+    return N && N->hasNUsesOfValue(NumUses, N.getResNo()) && P.match(DAG, N);
```
Comment (topperc) [inline llvm/include/llvm/CodeGen/SDPatternMatch.h:80]: hasNUsesOfValue can be expensive on nodes with multiple results. It can't return true until it has checked all uses. Uses for all results are stored in the same list.

We try to always check opcode first before checking number of uses. It's better to ask if an opcode is an ISD::ADD first since it only has 1 result. Than to ask for the number of uses without knowing what type of node it is.

Comment (mshockwave) [reply llvm/include/llvm/CodeGen/SDPatternMatch.h:80]: > We try to always check opcode first before checking number of uses. It's better to ask if an opcode is an ISD::ADD first since it only has 1 result. Than to ask for the number of uses without knowing what type of node it is.

I think a simple solution could be swapping `hasNUsesOfValue` with the succeeding `P.match` call, as we have to fulfills these two conditions anyway:
```
P.match(DAG, N) && N->hasNUsesOfValue(NumUses, N.getResNo())
```

Comment (mshockwave) [reply llvm/include/llvm/CodeGen/SDPatternMatch.h]: Done.

**Comment by RKSimon** [comment]
@mshockwave reverse-ping

**Comment by mshockwave** [comment]
> Random question - will this play nice with the template approach taken for VP node (e.g. #80105)?

My apologies for the delay.
I think supporting `VPMatchContext::match` in SDPatternMatch makes a lot of sense and I'll add that. Though I'm not sure if we should bring `VPMatchContext::getNode` into SDPatternMatch as well.

**Comment by mshockwave** [comment]
I'd just introduced the concept of match context to SDPatternMatch: now a user can pattern match a SDValue under certain match context using `sd_context_match`. A match context is really similar to that in DAGCombiner (e.g. VPMatchContext) except it only needs to implement `getDAG()` and `getTLI()`; The `match(SDValue, unsigned)` is optional. 
Furthermore, you can use custom contexts only in a sub-pattern via `m_Context(<custom context>, <sub pattern>)`.

Switching to this match context design also solves one of the problem where previously, the user had to provide a SelectionDAG in order to use any TLI hook, which was inconvenient in places lack of SelectionDAG. Now we can independently provide TLI instances to either `sd_match` or `sd_context_match`.

**Review by RKSimon — COMMENTED** [review]
Please can you add m_Deferred handling?

File: llvm/include/llvm/CodeGen/SDPatternMatch.h [hunk 2]
```diff
@@ -0,0 +1,712 @@
...
+      : DAG(nullptr), TLI(TLI) {}
+
+  // A valid MatchContext has to implement the following functions.
+
+  const SelectionDAG *getDAG() const { return DAG; }
+
+  const TargetLowering *getTLI() const {
+    if (TLI)
+      return TLI;
+    return DAG ? &DAG->getTargetLoweringInfo() : nullptr;
```
Comment (nikic) [inline llvm/include/llvm/CodeGen/SDPatternMatch.h:47]: I think either this should directly return TLI or the DAG initialization above should not fetch TLI.

File: llvm/include/llvm/CodeGen/SDPatternMatch.h [hunk 3]
```diff
@@ -0,0 +1,712 @@
...
+// === Utilities ===
+struct Value_match {
+  SDValue MatchVal;
+
+  Value_match() = default;
+
+  explicit Value_match(SDValue Match) : MatchVal(Match) {}
+
+  template <typename MatchContext> bool match(const MatchContext &, SDValue N) {
+    return (MatchVal && (MatchVal == N)) || N.getNode();
```
Comment (nikic) [inline llvm/include/llvm/CodeGen/SDPatternMatch.h:111]: I don't understand this condition. Won't this always match due to the `N.getNode()`, even if `MatchVal` is specified?

I think you should split this into separate matchers for m_Value and m_Specific.

File: llvm/include/llvm/CodeGen/SDPatternMatch.h [hunk 4]
```diff
@@ -0,0 +1,712 @@
...
+  const TargetLowering *getTLI() const {
+    if (TLI)
+      return TLI;
+    return DAG ? &DAG->getTargetLoweringInfo() : nullptr;
+  }
+
+  // Optional trait function(s)
+
+  /// Return true if N effectively has opcode Opcode.
+  // bool match(SDValue N, unsigned Opcode)
```
Comment (nikic) [inline llvm/include/llvm/CodeGen/SDPatternMatch.h:53]: Why make this optional instead of having a default implementation of `N->getOpcode() == Opcode`?

File: llvm/include/llvm/CodeGen/SDPatternMatch.h [hunk 5]
```diff
@@ -0,0 +1,712 @@
...
+
+struct Opcode_match {
+  unsigned Opcode;
+
+  explicit Opcode_match(unsigned Opc) : Opcode(Opc) {}
+
+  template <typename MatchContext>
+  std::enable_if_t<is_detected<ctx_has_match, MatchContext>::value, bool>
+  match(const MatchContext &Ctx, SDValue N) {
+    return N && Ctx.match(N, Opcode);
```
Comment (nikic) [inline llvm/include/llvm/CodeGen/SDPatternMatch.h:128]: Why the `if (N)` checks everywhere? I don't see why empty SDValues should be a valid input to matchers, and require checks everywhere.

File: llvm/include/llvm/CodeGen/SDPatternMatch.h [hunk 6]
```diff
@@ -0,0 +1,712 @@
...
+template <typename Pattern> struct TLI_pred_match {
+  Pattern P;
+  std::function<bool(const TargetLowering &, SDValue)> PredFunc;
+
+  TLI_pred_match(decltype(PredFunc) &&Pred, const Pattern &P)
+      : P(P), PredFunc(std::move(Pred)) {}
+
+  template <typename MatchContext>
+  bool match(const MatchContext &Ctx, SDValue N) {
+    return Ctx.getTLI() && N && PredFunc(*Ctx.getTLI(), N) && P.match(Ctx, N);
```
Comment (nikic) [inline llvm/include/llvm/CodeGen/SDPatternMatch.h:196]: The `Ctx.getTLI() &&` here seems like a footgun. It means the pattern will silently fail to match if TLI is not provided. I think we should assert instead.

File: llvm/include/llvm/CodeGen/SDPatternMatch.h [hunk 7]
```diff
@@ -0,0 +1,712 @@
...
+    BindVT = N.getValueType();
+    return true;
+  }
+};
+
+/// Retreive the ValueType of the current SDValue.
+inline ValueType_bind m_VT(EVT &VT) { return ValueType_bind(VT); }
+
+template <typename Pattern> struct ValueType_match {
+  std::function<bool(EVT)> PredFunc;
```
Comment (nikic) [inline llvm/include/llvm/CodeGen/SDPatternMatch.h:245]: Don't use std::function, it is super expensive. As you can't use function_ref here, I guess you'll have to template over the predicate as well.

File: llvm/include/llvm/CodeGen/SDPatternMatch.h [hunk 8]
```diff
@@ -0,0 +1,712 @@
...
+  Or(Pred &&p, Preds &&...preds)
+      : Or<Preds...>(std::forward<Preds>(preds)...), P(std::forward<Pred>(p)) {}
+
+  template <typename MatchContext>
+  bool match(const MatchContext &Ctx, SDValue N) {
+    return P.match(Ctx, N) || Or<Preds...>::match(Ctx, N);
+  }
+};
+
+template <typename... Preds> And<Preds...> m_all_of(Preds &&...preds) {
```
Comment (nikic) [inline llvm/include/llvm/CodeGen/SDPatternMatch.h:373]: Naming here is a bit inconsistent with everything else -- should be `m_AllOf`?

File: llvm/include/llvm/CodeGen/SDPatternMatch.h [hunk 9]
```diff
@@ -0,0 +1,712 @@
...
+
+    // This is the case where there are more predicates than operands.
+    return false;
+  }
+};
+
+template <typename... OpndPreds>
+Node_match<OpndPreds...> m_Node(unsigned Opcode, OpndPreds &&...preds) {
+  return Node_match<OpndPreds...>(Opcode, 0, std::forward<OpndPreds>(preds)...);
+}
```
Comment (nikic) [inline llvm/include/llvm/CodeGen/SDPatternMatch.h:438]: I didn't even try to understand this.

File: llvm/include/llvm/CodeGen/SDPatternMatch.h [hunk 10]
```diff
@@ -0,0 +1,712 @@
...
+      // and glue by default) and retreive the operand index offset.
+      EVT VT = N->getOperand(i).getValueType();
+      if (VT != MVT::Glue && VT != MVT::Other) {
+        ++Size;
+        if (FirstIndex == TotalNumOps)
+          FirstIndex = i;
+      }
+    }
+  }
+};
```
Comment (nikic) [inline llvm/include/llvm/CodeGen/SDPatternMatch.h:460]: This is going to impose a significant cost on all matches, even though the "effective operand" handling is usually not relevant (for stuff like ADDs, etc). I think this should be opt-in, or possibly removed entirely (match helpers can add wildcards at the appropriate positions).

File: llvm/include/llvm/CodeGen/SDPatternMatch.h [hunk 11]
```diff
@@ -0,0 +1,712 @@
...
+      : Opcode(Opc), LHS(L), RHS(R) {}
+
+  template <typename MatchContext>
+  bool match(const MatchContext &Ctx, SDValue N) {
+    if (!N)
+      return false;
+
+    if (sd_context_match(N, Ctx, m_Opc(Opcode))) {
+      EffectiveOperands EO(N);
+      if (EO.Size == 2)
```
Comment (nikic) [inline llvm/include/llvm/CodeGen/SDPatternMatch.h:479]: Should be an assert?

File: llvm/include/llvm/CodeGen/SDPatternMatch.h (hunk 4 above)
Comment (mshockwave) [reply llvm/include/llvm/CodeGen/SDPatternMatch.h:53]: Users won't necessary inherit BasicContext here to create a new match context. The default implementation for `match` is implemented in places like `Opcode_match::match`, toggled by `ctx_has_match`.

File: llvm/include/llvm/CodeGen/SDPatternMatch.h (hunk 8 above)
Comment (mshockwave) [reply llvm/include/llvm/CodeGen/SDPatternMatch.h:373]: This was inspired by `MIPatternMatch::m_all_of`, but I'll change to `m_AllOf`

File: llvm/include/llvm/CodeGen/SDPatternMatch.h (hunk 10 above)
Comment (mshockwave) [reply llvm/include/llvm/CodeGen/SDPatternMatch.h:440]: I choose to keep it and allow users to opt in via `m_ChainedBinOp`, as I still want to match some binary operations like `STRICT_F*` with this.

File: llvm/include/llvm/CodeGen/SDPatternMatch.h (hunk 9 above)
Comment (mshockwave) [reply llvm/include/llvm/CodeGen/SDPatternMatch.h:418]: This is trying to generically match a node by its opcode and operands using variadic template argument tricks. The implementation is similar to `m_AllOf` and `m_AnyOf` above.

**Comment by mshockwave** [comment]
> Please can you add m_Deferred handling?

Of course. I'll add it in the next update.

File: llvm/include/llvm/CodeGen/SDPatternMatch.h (hunk 11 above)
Comment (mshockwave) [reply llvm/include/llvm/CodeGen/SDPatternMatch.h]: Done.

File: llvm/include/llvm/CodeGen/SDPatternMatch.h (hunk 7 above)
Comment (mshockwave) [reply llvm/include/llvm/CodeGen/SDPatternMatch.h]: Done.

File: llvm/include/llvm/CodeGen/SDPatternMatch.h (hunk 6 above)
Comment (mshockwave) [reply llvm/include/llvm/CodeGen/SDPatternMatch.h]: Done.

File: llvm/include/llvm/CodeGen/SDPatternMatch.h (hunk 5 above)
Comment (mshockwave) [reply llvm/include/llvm/CodeGen/SDPatternMatch.h]: Done.

File: llvm/include/llvm/CodeGen/SDPatternMatch.h (hunk 3 above)
Comment (mshockwave) [reply llvm/include/llvm/CodeGen/SDPatternMatch.h]: It's fixed now.

File: llvm/include/llvm/CodeGen/SDPatternMatch.h (hunk 2 above)
Comment (mshockwave) [reply llvm/include/llvm/CodeGen/SDPatternMatch.h]: Done.

File: llvm/include/llvm/CodeGen/SDPatternMatch.h (hunk 4 above)
Comment (nikic) [reply llvm/include/llvm/CodeGen/SDPatternMatch.h:49]: Do we expect there to be more contexts than BaseContext and VPMatchContext? I don't think that custom context will be common enough that we have to optimize for implementing one method less.

File: llvm/include/llvm/CodeGen/SDPatternMatch.h [hunk 12]
```diff
@@ -0,0 +1,731 @@
...
+/// Provide number of operands that are not chain or glue, as well as the first
+/// index of such operand.
+template <bool ExcludeChain> struct EffectiveOperands {
+  unsigned Size = 0;
+  unsigned FirstIndex = 0;
+
+  explicit EffectiveOperands(SDValue N) {
+    const unsigned TotalNumOps = N->getNumOperands();
+    FirstIndex = TotalNumOps;
+    for (unsigned i = 0; i < TotalNumOps; ++i) {
```
Comment (nikic) [inline llvm/include/llvm/CodeGen/SDPatternMatch.h:451]: i -> I

File: llvm/include/llvm/CodeGen/SDPatternMatch.h [hunk 13]
```diff
@@ -0,0 +1,731 @@
...
+    BindVal = N;
+    return true;
+  }
+};
+
+inline Value_bind m_Value(SDValue &N) { return Value_bind(N); }
+
+template <typename Pattern> struct TLI_pred_match {
+  Pattern P;
+  std::function<bool(const TargetLowering &, SDValue)> PredFunc;
```
Comment (nikic) [inline llvm/include/llvm/CodeGen/SDPatternMatch.h:206]: Same here, avoid std::function.

File: llvm/include/llvm/CodeGen/SDPatternMatch.h (hunk 9 above)
Comment (nikic) [reply llvm/include/llvm/CodeGen/SDPatternMatch.h:440]: I find the implementation of this one quite hard to follow, compared to m_AllOg/m_AnyOf. Especially the presence of Opcode and OpIndex as members in each recursively inherited struct seems odd. OpIndex should probably be a template parameter, and Opcode checked separately from the rest.

File: llvm/include/llvm/CodeGen/SDPatternMatch.h (hunk 12 above)
Comment (mshockwave) [reply llvm/include/llvm/CodeGen/SDPatternMatch.h]: Done

File: llvm/include/llvm/CodeGen/SDPatternMatch.h (hunk 13 above)
Comment (mshockwave) [reply llvm/include/llvm/CodeGen/SDPatternMatch.h]: Done

File: llvm/include/llvm/CodeGen/SDPatternMatch.h (hunk 9 above)
Comment (mshockwave) [reply llvm/include/llvm/CodeGen/SDPatternMatch.h:425]: > I find the implementation of this one quite hard to follow, compared to m_AllOg/m_AnyOf. Especially the presence of Opcode and OpIndex as members in each recursively inherited struct seems odd. OpIndex should probably be a template parameter, and Opcode checked separately from the rest.

I've simplified this by only doing recursive template matching on operands and check the opcode in `m_Node`.

File: llvm/include/llvm/CodeGen/SDPatternMatch.h (hunk 4 above)
Comment (mshockwave) [reply llvm/include/llvm/CodeGen/SDPatternMatch.h:49]: > Do we expect there to be more contexts than BaseContext and VPMatchContext? I don't think that custom context will be common enough that we have to optimize for implementing one method less.

I'm convinced, I'll change this in the next update.

Comment (mshockwave) [reply llvm/include/llvm/CodeGen/SDPatternMatch.h]: It's fixed now.

File: llvm/include/llvm/CodeGen/SDPatternMatch.h [hunk 14]
```diff
@@ -0,0 +1,716 @@
...
+  return DeferredValue_match(V);
+}
+
+struct Opcode_match {
+  unsigned Opcode;
+
+  explicit Opcode_match(unsigned Opc) : Opcode(Opc) {}
+
+  template <typename MatchContext>
+  std::enable_if_t<is_detected<ctx_has_match, MatchContext>::value, bool>
```
Comment (nikic) [inline llvm/include/llvm/CodeGen/SDPatternMatch.h:146]: I think these enable_ifs are not needed anymore now that we have the match() implementation in BasicMatchContext?

Comment (mshockwave) [reply llvm/include/llvm/CodeGen/SDPatternMatch.h]: Oh yeah I forgot. It's fixed now.

**Review by nikic — APPROVED** [review]
LGTM, but please wait a bit in case other people have more feedback.
